### PR TITLE
Clean build against nightly-2023-04-14

### PR DESCRIPTION
**Pull request by thomasjm**
Similar to #1629, I'm opening this to see what the CI says.

### CODE REVIEW

File: stack.yaml [hunk 1]
```diff
@@ -17,4 +18,5 @@ packages:
 # - doc/tutorial/
 
 extra-deps:
-- hspec-wai-0.10.1
+- hspec-wai-0.11.1
+- pipes-safe-2.3.4
```
Comment (ysangkok) [inline stack.yaml:22]: Possibly unnecessary soon:

* commercialhaskell/stackage/pull/6943

Comment (thomasjm) [reply stack.yaml:22]: Any chance you could hit the CI workflow button?

Comment (ysangkok) [reply stack.yaml:22]: I can't, I don't maintain Servant

Comment (ysangkok) [reply stack.yaml:22]: @tchoutri Would you be able to approve the workflow?

**Comment by thomasjm** [comment]
Okay, I added an hspec constraint for GHCJS and the tests are passing now: https://github.com/codedownio/servant/actions/runs/4726572557. 

In summary this PR does the following:

* Bump `hspec` bounds for everything except GHCJS, so the latest Stackage nightly works (GHC 9.4.4)
* Modernize the CI workflow a little bit, including bumping GHC minor versions
* Update the `tested-with` field in the cabal files

So I think this PR is ready to go and can supersede #1629

**Comment by ysangkok** [comment]
@tchoutri Looks good to me! Did something come up?

**Comment by tchoutri** [comment]
It did! Let's merge :)

**Comment by thomasjm** [comment]
Thanks all! I think we can close #1623 and #1629 now

**Pull request merged**